### PR TITLE
Gh-2779: FederatedStore remove handlers on RemoveGraph

### LINKED ISSUE #2779: FederatedStore FederatedAddGraphHandlerParent appending supported operations to FederatedStore
Within the `FederatedAddGraphHandlerParent` there is logic to add Handlers for operations that are unsupported by the FederatedStore, WHEN adding a new graph that DOES support the new operation, which the FederatedStore did not previously support.

This Logic should be explored to confirm that it is still required.

THEN it likely in `RemoveGraph` when the LAST graph which supports an operation is removed, it should remove the supported operations from the FederatedStore.

Additional info
The handlers which are added this way to support these new operations are just hollow gateway handlers which typically just wrap the operation in a `FederatedOperation`. Attention should  be paid to if the `AdminConfiguredMergeMap` requires some kind of updating within the FederatedStore constructor.

### PR DESCRIPTION
**Pull request by t92549**
# Related Issue

- Resolve #2779

### CODE REVIEW

**Comment by t92549** [comment]
This implementation is a bit messy, but necessary without a large rework to how operation handlers and supported operations are implemented by the FederatedStore.

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java [hunk 1]
```diff
@@ -35,10 +40,34 @@
...
+                        store.addOperationHandler(removedOperation, null);
+                        ((FederatedStore) store).removeExternallySupportedOperation(removedOperation);
+                    }
+                }
+            }
+        }
+        return removed;
+    }
+
+    private Set<Class<? extends Operation>> getSupportedOperations(final Context context, final FederatedStore store) {
```
Comment (GCHQDev404) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:65]:
```suggestion
    private Set<Class<? extends Operation>> getSupportedOperationsFromSubGraphs(final Context context, final FederatedStore store) {
```

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java [hunk 2]
```diff
@@ -35,10 +40,34 @@
...
+            removed = ((FederatedStore) store).remove(operation.getGraphId(), context.getUser(), operation.isUserRequestingAdminUsage());
         } catch (final Exception e) {
             throw new OperationException("Error removing graph: " + operation.getGraphId(), e);
         }
+        if (removed) {
+            Set<Class<? extends Operation>> updatedSupportedOperations = getSupportedOperations(context, (FederatedStore) store);
+            if (!supportedOperations.equals(updatedSupportedOperations)) {
+                supportedOperations.removeAll(updatedSupportedOperations);
+                for (final Class<? extends Operation> removedOperation : supportedOperations) {
+                    if (((FederatedStore) store).getExternallySupportedOperations().contains(removedOperation)) {
```
Comment (GCHQDev404) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:55]: else? what happens?
is that illegal state? do we care?

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java [hunk 3]
```diff
@@ -135,6 +135,7 @@ public class FederatedStore extends Store {
     private List<String> storeConfiguredGraphIds;
     @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "class")
     private Map<String, BiFunction> storeConfiguredMergeFunctions;
+    private Set<Class<? extends Operation>> externallySupportedOperations = new HashSet<>();
```
Comment (GCHQDev404) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java:138]: You can delete this ;) Its not used anywhere.
You set and get it, but it has not effect on `graph.getSupported operations()` 

you check for the diff before and after removing a graph. you update `supportedOperations` and you also update this superfluous object which isn't used anywhere.

**Review by GCHQDev404 — CHANGES_REQUESTED** [review]

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java (hunk 2 above)
Comment (t92549) [reply store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:55]: The if statement is needed because it checks that the operations that are no longer supported by subgraphs are not operations that should still be supported by the federated store itself.
Otherwise if you have 2 subgraphs, an Accumulo store and a federated store, and you removed the sub federated store, all of the federated operation handlers would be deleted. This is why the `getExternallySupportedOperations` flag is required.

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java (hunk 3 above)
Comment (t92549) [reply store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java:138]: See my other comment below, this is required

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java (hunk 2 above)
Comment (t92549) [reply store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:55]: So to answer "else what happens?": If one store is added and then removed, the `GetElementsHandler` would be detected as removed from the sub graphs, but we see that this is not in the `externallySupportedOperations` as the federated store can support this itself, so we do not delete the handler.

Comment (GCHQDev404) [reply store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:55]: Can you review this logic.

1) AddGraph(FedStore) // has AddGraph
2) FederatedAddGraphParentHandler -> addExternallySupportedOperations() // AddGraph is added.
3) RemoveGraphHandler line44 -> getSupportedOperations // includes AddGraph from (1)
4) RemoveGraphHandler line51 -> getSupportedOperations // does NOT include AddGraph
5) line 52 not equal
6) line 53 supported list has AddGraph (it wasn't removed)
7) line 55 externallySupportedOperations contains AddGraph 
8) line 56 remove AddGraph from current FederatedStore.

FederatedStore no supports AddGraphs.

Comment (t92549) [reply store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:55]: See new test `shouldKeepFederatedStoreSupportedOperations`

Comment (t92549) [reply store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:55]: `AddGraph` is never added to `externallySupportedOperations` because in the [`FederatedAddGraphHandlerParent::addGenericHandler`](https://github.com/gchq/Gaffer/blob/dce22c074b9fcc89cb6f4b2c9cc8e193217e8066/store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/FederatedAddGraphHandlerParent.java#L94) it only gets added when `!store.isSupported(supportedOperation)`

Comment (GCHQDev404) [reply store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedRemoveGraphHandler.java:55]: thanks. I think I'm satisfied by this.

**Review by GCHQDev404 — APPROVED** [review]

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java [hunk 4]
```diff
@@ -580,7 +581,7 @@ public boolean isSupported(final Class<? extends Operation> operationClass) {
      * @return a collection of all the supported {@link Operation}s.
      */
     public Set<Class<? extends Operation>> getSupportedOperations() {
-        return operationHandlers.keySet();
+        return new HashSet<>(operationHandlers.keySet());
```
Comment (GCHQDeveloper314) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java:584]: Why is this needed? It seems unrelated to this PR.

**Review by GCHQDeveloper314 — APPROVED** [review]

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java (hunk 4 above)
Comment (t92549) [reply core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java:584]: I wanted to safely return a copy of the keys so I could modify the result without modifying the underlying operationHandlers map
